pkg/clusteragent/clusterchecks/dispatcher_rebalance.go

```diff
@@ -0,0 +1,208 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// +build clusterchecks
+
+package clusterchecks
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type Weight struct {
+	nodeName string
+	busyness int
+}
+
+type Weights []Weight
+
+func (w Weights) Len() int           { return len(w) }
+func (w Weights) Less(i, j int) bool { return w[i].busyness > w[j].busyness }
+func (w Weights) Swap(i, j int)      { w[i], w[j] = w[j], w[i] }
+
+func (d *dispatcher) calculateAvg() (int, error) {
+	busyness := 0.0
+	length := 0
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for _, node := range d.store.nodes {
+		node.RLock()
+		for _, stats := range node.clcRunnerStats {
+			busyness += busynessFunc(stats.AverageExecutionTime, stats.MetricSamples)
+		}
+		length += 1
+		node.RUnlock()
+	}
+
+	if length == 0 {
+		return -1, fmt.Errorf("zero nodes reporting")
+	}
+
+	return int(busyness) / length, nil
+}
+
+// getDiffAndWeights creates a map that contains the difference between
+// the busyness on each node and the total average busyness, and a Weights
+// struct containing nodes and their busyness values
+func (d *dispatcher) getDiffAndWeights(avg int) (map[string]int, Weights) {
+	diffMap := make(map[string]int)
+	weights := Weights{}
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for nodeName, node := range d.store.nodes {
+		node.RLock()
+		busyness := calculateBusyness(node.clcRunnerStats)
```

Comment:
to many lock... :)
what to you think about doing something like:
```golang
func (node *Node) GetBusyness(busynessFunc func(avgExecTime, mSamples int) float64) int {
        busyness := 0.0
        node.stats.RLock()
        defer node.stats.RUnLock()
	for _, stats := range checkStats {
               busyness += busynessFunc(stats.AverageExecutionTime, stats.MetricSamples)
	}
	return int(busyness)
}
```

It will remove the fact you need to `Lock()` and `Unlock()` a struct for outside, in several places

Reply:
I agree this would make things simpler, but I didn't want to break the locking pattern we have already: `// Lock is to be held by the user (dispatcher)`
wdyt? should we make the exception this time?